File: R/cyclical-GP/exponential.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(x);
  DATA_VECTOR(y);
  PARAMETER(b0);
  PARAMETER(log_gp_sigma);
  PARAMETER(log_gp_theta);
  PARAMETER(log_sigma);
  PARAMETER_VECTOR(eta)
  DATA_INTEGER(flag);
  
  int n = y.size(); 
  
  Type gp_sigma = exp(log_gp_sigma);
  Type gp_theta = exp(log_gp_theta);
  Type sigma = exp(log_sigma);
  REPORT(sigma);
  REPORT(gp_sigma);
  REPORT(gp_theta);
  
  matrix<Type> C(n, n);
  C.setZero();  
  for(int i=0; i<C.rows(); i++)
    for(int j=0; j<C.cols(); j++) {
      Type dist_sq = (x[j] - x[i]) * (x[j] - x[i]);
      C(i, j) = gp_sigma*gp_sigma * exp(- 1 / (2 * gp_theta * gp_theta) * dist_sq);
    }
    
  Type jnll = density::MVNORM(C)(eta);
  if (flag == 0) return jnll;
  jnll-= dnorm(y, vector<Type>(b0 + eta), sigma, true).sum();
  
  jnll-= dnorm(gp_sigma, Type(0), Type(2));
  jnll-= dnorm(gp_theta, Type(0), Type(3));
  jnll-= dnorm(b0, Type(0), Type(10));
  jnll-= dnorm(sigma, Type(0), Type(1));
  jnll-= log_gp_sigma;
  jnll-= log_gp_theta;
  jnll-= log_sigma;
  
  return jnll;
}

File: R/cyclical-GP/exponential.R
# GP regression example

GP_sq_exp <- function(x, b0 = 0, sigma = 0.1, gp_sigma = 0.1, gp_theta = 2) {
  Sigma <- sapply(x, function(x1) {
    sapply(x, function(x2) {
      dist_sq <- (x2 - x1)^2
      gp_sigma*gp_sigma * exp(- 1 / (2 * gp_theta * gp_theta) * dist_sq)
    })
  })
  eta <- MASS::mvrnorm(mu = rep(0, length(x)), Sigma = Sigma)
  y <- rnorm(length(x), mean = b0 + eta, sd = sigma)
  list(x = x, y = y, eta = eta)
}

library(TMB)
compile("exponential.cpp")
dyn.load(dynlib("exponential"))

set.seed(108)
out <- GP_sq_exp(1:40, b0 = 0, sigma = 0.2, gp_sigma = 0.4, gp_theta = 1.5)
plot(out$x, out$eta, type = "l")
points(out$x, out$y)

data <- list(x = out$x, y = out$y, flag = 1)
parameters <- list(
  b0 = 0,
  log_gp_sigma = log(0.5), 
  log_gp_theta = log(1),
  log_sigma = log(0.1), 
  eta = rep(0, length(out$x))
)
obj <- TMB::MakeADFun(data, parameters, DLL="exponential",
  random = "eta")
obj <- normalize(obj, flag="flag")
opt <- stats::nlminb(obj$par, obj$fn, obj$gr)
sdr <- TMB::sdreport(obj)
sdr

r <- obj$report()
print(r)

